src/main/php/peer/server/AsyncServer.class.php

```diff
@@ -0,0 +1,205 @@
+<?php namespace peer\server;
+
+use lang\{IllegalStateException, Throwable};
+use peer\server\protocol\SocketAcceptHandler;
+use peer\{ServerSocket, SocketTimeoutException};
+
+/**
+ * Asynchronous TCP/IP Server
+ *
+ * ```php
+ * use peer\server\AsyncServer;
+ *   
+ * $server= new AsyncServer();
+ * $server->listen(new ServerSocket('127.0.0.1', 6100), new MyProtocol());
+ * $server->service();
+ * $server->shutdown();
+ * ```
+ *
+ * @see   xp://peer.ServerSocket
+ * @test  xp://peer.unittest.server.AsyncServerTest
+ */
+class AsyncServer extends Server {
+  private $select= [], $handle= [], $tasks= [];
+
+  /**
+   * Adds server socket to listen on, associating protocol handler with it
+   *
+   * @param  peer.ServerSocket|peer.BSDServerSocket $socket
+   * @param  peer.server.ServerProtocol $protocol
+   * @return self
+   */
+  public function listen($socket, $protocol) {
+    $protocol->server= $this;
+
+    $socket->create();
+    $socket->bind(true);
+    $socket->listen();
+
+    $this->select[]= $socket;
+    $this->handle[]= [function($socket) use($protocol) {
+      $connection= $socket->accept();
+      if ($protocol instanceof SocketAcceptHandler && !$protocol->handleAccept($connection)) {
+        $connection->close();
+        return;
+      }
+
+      $this->tcpnodelay && $connection->useNoDelay();
+      $protocol->handleConnect($connection);
+
+      $this->select[]= $connection;
+      $this->handle[]= [
+        [$protocol, 'handleData'],
+        [$protocol, 'handleDisconnect'],
+        [$protocol, 'handleError'],
+        [$protocol, 'initialize']
+      ];
```

Comment:
"Isn't this slow?" - I thought - indirectly invoking methods via an array?

```php
use peer\Socket;
use util\profiling\Timer;

class Protocol {

  public function handleData($socket) {
    return $socket->toString();
  }
}

$protocol= new Protocol();
$invoke= [[$protocol, 'handleData']];
$socket= new Socket('127.0.0.1', 8080);

$timer= (new Timer())->start();
for ($i= 0; $i < 10_000_000; $i++) {
  // $result= ($f= $invoke[0] ?? null) ? $f($socket) : null;
  $result= $protocol->handleData($socket);
}

printf("%.3f seconds: %s", $timer->elapsedTime(), $result);
```

* Indirect: 5.791 seconds: peer.Socket((closed) -> tcp://127.0.0.1:8080)
* Direct: 4.877 seconds: peer.Socket((closed) -> tcp://127.0.0.1:8080)

So yes, this is one second difference, but for 10 million invocations (*so only 1_000 / 10_000_000 = 0.0001 milliseconds difference per invocation*)! Compared to the socket I/O we're handling